specs/makeStyles.md

```diff
@@ -0,0 +1,322 @@
+Parent component passes its overrides to a child as a list of classes in `className` prop. The child, takes the input classes and references the global dictionary to get back the necessary CSS attribute names to be able to perform the merge - the merge is implemented in `ax()` function as described in #16411.
+
+#### Alternative 1: encode the CSS property name to the hash
+
+Instead of using the global dictionary side effect, we can use part of the hash in the classname to encode the CSS property:
+![](makeStyles-classname.png)
+
+Then when merging in a child, we can merge the classes by replacing previous classnames with the same CSS property hash.
+This approach makes the classnames longer and according to the test does not make the merging any faster.
+
+#### Alternative 2: use special `styles` prop for overrides
+
+If the global dictionary does not perform as expected, we might return back to the previous v0 approach where we use a dedicated `styles` prop to pass overrides from parent to child.
+
+### Another libraries using similar approach
+
+This has been inspired by [Facebook stylex talk](https://www.youtube.com/watch?v=9JZHodNR184).
+Similar approach is also used in Attlasian [compiled v5](https://github.com/atlassian-labs/compiled).
+
+## Sample Code
+
+```jsx
+// Button.tsx
+const useButtonStyles = makeStyles([
+  [null, { cursor: 'pointer' }], // always applied
+  [props => props.primary, tokens => ({ backgroundColor: tokens.colors.primary.background })],
+  [props => pros.disabled, { backgroundColor: 'gray' }],
+]);
+
+const Button = ({ className, ...props }) => {
+  const buttonClasses = useButtonStyles(props); // button styles
+  const mergedClasses = ax(buttonClasses, className); // merge with `className` = overrides from parent
+  return <button className={mergedClasses} {...props} />;
+};
+
+// ButtonUsage.tsx
+const useButtonOverrides = makeStyles([[null, { backgroundColor: 'red' }]]);
+
+const ButtonUsage = () => {
+  const buttonOverrides = useButtonOverrides({});
+
+  return <Button classname={buttonOverrides}>I have always red background</Button>;
+};
+```
+
+## Issues, open questions
+
+### Theme-level style overrides
+
+As described in _Applying theme to styles_, a theme can only define tokens but can never change styles. There is a concern that this is not flexible enough and there might be a valid scenario where theme-specific style overrides will be required.
+How would that be implemented.
+Example API:
+
+```js
+// The Button styles need an id to pull overrides
+// from the theme:
+const makeStyles({ id: 'Button', { ... } });
+
+// The theme needs to be able to define them:
+const theme = {
+  styles: {
+    Button: { // id?
+      styles: makeStyles({ ... }) // maybe?
+    }
+  }
+}
+```
+
+### Always matching matcher
+
+What is the best signature for a `makeStyles` call with a single style object which always matches?
+
+```js
+makeStyles([[null, { background: 'red ' }]]);
+makeStyles([{ background: 'red' }]);
```

Comment:
**thought:** 
- `[[null, {stylesDef}]]`: explicit need to use `null` for positional args, feels rather as an anti-pattern. also this isn't very "user friendly" syntax
- `[{stylesDef}]`: need of wrapping pure object into an array feels kinda strange, on the other hand makes it more consistent (also can be explicitly typed / enforce that only first array item can contain non tuple)
- `{stylesDef}`: cleanest way, although will get messy when dynamic rules need to be applied

**Q (probably out off topic)**

Is there any particular reason why we cannot embrace "common cssInJS APIs for defining dynamic definitions", like  following ( I assume ability to process rules in compile step ? )
```
makeStyles((props:ButtonProps) => ({
   // static definitions
   color: 'red';
   '&:focus,&:hover': { ... },
   // dynamic defs
   backgroundColor: props.primary ?? tokens.colors.primary.background
}))
```

Reply:
**q:** What is expect to happen if user provides multiple "apply all rules" ?

```
  [null, { cursor: 'pointer' }], // always applied
  [props => props.primary, tokens => ({ backgroundColor: tokens.colors.primary.background })]
  [null, { color: 'red' }], // always applied
```

I assume it will be merged right? do we wanna allow this from API standpoint ( if not , my previous comment regarding `[{stylesDef}]` would be great candidate for "always applied styles" )

Reply:
> Is there any particular reason why we cannot embrace "common cssInJS APIs for defining dynamic definitions", like following ( I assume ability to process rules in compile step ? )

`backgroundColor: props.primary ?? tokens.colors.primary.background` there is no way to make this build time compatible, even if we will replace `props.primary`, for example, it will not produce the same thing.

[Linaria](https://github.com/callstack/linaria) and [Compiled](https://compiledcssinjs.com/) have similar restrictions.